POEM_035.md

```diff
@@ -0,0 +1,44 @@
+POEM ID: 035  
+Title: More generalized behavior in promoted inputs.  
+authors: [robfalck]  
+Competing POEMs: N/A  
+Related POEMs: N/A  
+Associated implementation PR:
+
+##  Status
+
+- [x] Active
+- [ ] Requesting decision
+- [ ] Accepted
+- [ ] Rejected
+- [ ] Integrated
+
+
+## Motivation
+
+The current input promotion system has exposed an inconsistency that can cause confusion among users and makes certain connections impossible.
+
+## Description
+
+Currently, if a variable is promoted using src_indices and it receives no connection, an automatic IndepVarComp output is created for it using those source indices to infer its shape.
+However, if an output is connected to it with some set of source indices, it currently generates an error.
+The `src_indices` from the connect call and the promotes call may conflict, and this is currently an error.
+This is a weakness in the current API.  A complex model may utilize src_indices deep within a nested system.
+A user may unwittingly try to connect to that with `src_indices` and encounter an error.
+
+![Current behavior for sizing a promoted input](POEM_035/current_behavior.png)
+
+Instead, this POEM proposes making src_indices apply to each level of promotion.
+During setup, OpenMDAO can then traverse the chain of connections and resolve which elements of the ultimate source end up where in the ultimate target.
+
+### Proposals
+
+1. `src_shape` is added as an argument to `promotes` and `add_subsystem` (when inputs are promoted), which specifies the apparent shape of the promoted input (if connecting to it) or the related AutoIVC output (if left unconnected).
```

Comment:
I don't see how we can make it work as an argument to `add_subsystem`. In the other cases (`set_input_defaults`, and `promotes` you have the granularity to break out specific inputs or subsets of inputs that share common shape and src_indices. In `add_subsystem` you don't. 

We don't currently allow src_inidices in `add_subsystem` and I don't think we should allow `src_shape` there either. I think that using the promotes arguments for `add_subsystem` should just be a slightly less capable API

Reply:
Allow `src_shape` to be a dict.

```
self.add_subsystem('foo', promotes_inputs=['a', ('b', 'b_in'), 'c'],
                                    src_indices={'b_in': src_idxs},
                                    src_shape={'b_in': (4, 3))
```